control-plane/cni/main.go

```diff
@@ -135,56 +162,119 @@ func cmdAdd(args *skel.CmdArgs) error {
+		logger.Error("could not apply iptables setup: %v", err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
 
-	// Pass through the result for the next plugin
+	logger.Debug("traffic redirect rules applied to pod: %s", pod.Name)
+	// Pass through the result for the next plugin even though we are the final plugin in the chain.
 	return types.PrintResult(result, cfg.CNIVersion)
 }
 
 // cmdDel is called for DELETE requests.
 func cmdDel(args *skel.CmdArgs) error {
-	// Nothing for consul-cni plugin to do as everything is removed once the pod is gone
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
+// cmdCheck is called for CHECK requests.
 func cmdCheck(args *skel.CmdArgs) error {
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
 func main() {
 	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.All, bv.BuildString("consul-cni"))
 }
 
-func hasBeenInjected(pod corev1.Pod) bool {
-	if anno, ok := pod.Annotations[keyInjectStatus]; ok && anno == injected {
+// skipConnectInject determines if the connect-inject annotation is false.
+func skipConnectInject(pod corev1.Pod) bool {
+	if anno, ok := pod.Annotations[annotationInject]; ok && anno == "false" {
 		return true
 	}
 	return false
 }
+
+// waitForAnnotation waits for an annotation to be available. Returns immediately if the annotation exists.
+func waitForAnnotation(pod corev1.Pod, annotation string, retries uint64) bool {
+	var err error
+	err = backoff.Retry(func() error {
+		var ok bool
+		_, ok = pod.Annotations[annotation]
+		if !ok {
+			return fmt.Errorf("annotation %s does not exist yet", annotation)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), retries))
+	return err == nil
+}
+
+// parseAnnotation parses the cni-proxy-config annotation into an iptables.Config object.
+func parseAnnotation(pod corev1.Pod, annotation string) (iptables.Config, error) {
+	anno, ok := pod.Annotations[annotation]
+	if !ok {
+		return iptables.Config{}, fmt.Errorf("could not find %s annotation for %s pod", annotation, pod.Name)
+	}
+	cfg := iptables.Config{}
+	err := json.Unmarshal([]byte(anno), &cfg)
+	if err != nil {
+		return iptables.Config{}, fmt.Errorf("could not unmarshal %s annotation for %s pod", annotation, pod.Name)
+	}
+	return cfg, nil
+}
+
+// searchDNSIPFromEnvironment gets the consul server DNS IP from the pods environment variables. The prefix makes searching easier
+// return an empty string.
+func searchDNSIPFromEnvironment(pod corev1.Pod, prefix string) string {
+	var result string
+	upcaseResourcePrefix := strings.ToUpper(prefix)
+	upcaseResourcePrefixWithUnderscores := strings.ReplaceAll(upcaseResourcePrefix, "-", "_")
+	dnsName := strings.Join([]string{upcaseResourcePrefixWithUnderscores, dnsServiceHostEnvSuffix}, "_")
+
+	// Environment variables are buried in the pod spec
+	vars := pod.Spec.Containers[0].Env
```

Comment:
This is a Pod we control, right? I'm not 100% up to speed on CNI, but because this is a Pod that we have full control over, we are safe making the assumption that this container is always first in the spec?

Reply:
It is our pod yes but environment variables are for an entire pod. It wouldn't have mattered if I used Containers[1] or Containers[2].

Reply:
That makes sense. I hadn't considered that of course the environment variables are the same across all of the containers in a Pod, but it's obvious when you say it.